File: 12. Partition Equal Subset Sum.cpp
// 12. Partition Equal Subset Sum
// Topic: Partition Equal Subset Sum
// Description: Check if array can be partitioned into two subsets with equal sum using memoized
// subset sum. Input: n followed by n integers Output: 1 (true) or 0 (false)

#include <bits/stdc++.h>
using namespace std;

int canPartitionFrom(int idx, int sum, const vector<int>& arr, vector<vector<int>>& memo)
{
    if (sum == 0)
        return 1;
    if (idx >= (int)arr.size() || sum < 0)
        return 0;
    if (memo[idx][sum] != -1)
        return memo[idx][sum];
    int take = canPartitionFrom(idx + 1, sum - arr[idx], arr, memo);
    int skip = canPartitionFrom(idx + 1, sum, arr, memo);
    memo[idx][sum] = (take || skip);
    return memo[idx][sum];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<int> arr(n);
    int total = 0;
    for (int i = 0; i < n; ++i)
    {
        cin >> arr[i];
        total += arr[i];
    }
    if (total % 2 == 1)
    {
        cout << 0 << '\n';
        return 0;
    }
    int target = total / 2;
    vector<vector<int>> memo(n, vector<int>(target + 1, -1));
    cout << canPartitionFrom(0, target, arr, memo) << '\n';
}

// Memoization process started in canPartitionFrom where memo caches subproblem results.
// Time Complexity: O(n * target)
// Space Complexity: O(n * target)

File: 12. Partition Equal Subset Sum.py
# 12. Partition Equal Subset Sum
# Topic: Partition Equal Subset Sum
# Description: Determine if the array can be partitioned into two subsets with equal sum using memoized subset-sum.
# Input: n followed by n integers
# Output: 1 (true) or 0 (false)

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    arr = [int(next(it)) for _ in range(n)]
    total = sum(arr)
    if total % 2 != 0:
        print(0)
        sys.exit(0)
    target = total // 2
    arr = [abs(x) for x in arr]

    @lru_cache(maxsize=None)
    def can(i: int, rem: int) -> int:
        if rem == 0:
            return 1
        if i >= n or rem < 0:
            return 0
        return 1 if (can(i+1, rem - arr[i]) or can(i+1, rem)) else 0

    print(can(0, target))

# Memoization process started at can decorated with lru_cache.
# Time Complexity: O(n * target)
# Space Complexity: O(n * target)